File: python/28_sense_and_move_2.py
# Modify the move function to accommodate the added
# probabilities of overshooting or undershooting
# the intended destination.
from typing import Sequence

p = [0.2, 0.2, 0.2, 0.2, 0.2]
world = ['green', 'red', 'red', 'green', 'green']
measurements = ['red', 'red']
pHit = 0.6
pMiss = 0.2

motions = [1, 1]
pExact = 0.8
pOvershoot = 0.1
pUndershoot = 0.1


def sense(p: Sequence[float], Z: float) -> Sequence[float]:
    q = []
    for i in range(len(p)):
        hit = (Z == world[i])
        q.append(p[i] * (hit * pHit + (1 - hit) * pMiss))
    s = sum(q)
    for i in range(len(q)):
        q[i] = q[i] / s
    return q


def move(p: Sequence[float], U: float) -> Sequence[float]:
    n = len(p)
    q = []
    for i in range(len(p)):
                                     # i is index of cell we are moving to
        under = (i - (U + 1)) % n    # index of cell where undershoot from the cell would move us to i
        exact = (i - U) % n          # index of cell where exact move gets us to i (centroid of where we moved from)
        over = (i - (U - 1)) % n     # index of cell where overshoot from the cell would move us to i

        # sum probabilities of all the movements that could have moved us to i
        q.append(p[under] * pUndershoot + p[exact] * pExact + p[over] * pOvershoot)
    return q

q = p
for i in range(len(motions)):
    q = move(sense(q, measurements[i]), motions[i])
print(q)

#
# Test float equality within a tolerance
#
def is_approximately(x: float, y: float, tolerance: float) -> bool:
    return abs(x - y) <= abs(tolerance)

#
# test list of float equality with tolerance
#
def are_approximately(x: Sequence[float], y: Sequence[float], tolerance: float) -> bool:
    if len(x) != len(y):
        return False
    for i in range(len(x)):
        if not is_approximately(x[i], y[i], tolerance):
            return False
    return True

assert(is_approximately(sum(q), 1.0, 0.0001))
assert(are_approximately(q, [0.0788, 0.0753, 0.2247, 0.4329, 0.1882], 0.0001))

File: swift/28_sense_and_move_2.swift
// Program a function that returns a new distribution
// q, shifted to the right by U units. If U=0, q should
// be the same as p.

var p = [0.2, 0.2, 0.2, 0.2, 0.2]
let world = ["green", "red", "red", "green", "green"]
let measurements = ["red", "red"]
let pHit = 0.6
let pMiss = 0.2

let motions = [1, 1]
let pExact = 0.8
let pOvershoot = 0.1
let pUndershoot = 0.1

/*
** apply measurement the prior and return the posterior
*/
func sense(_ prior: [Double], _ measurement: String) -> [Double] {
    //
    // ADD YOUR CODE HERE
    // 
    let q = prior.enumerated().map {(index, value) in value * ((measurement == world[index]) ? pHit : pMiss) }
    let sum = q.sum()
    return q.map {$0 / sum}
}

func move(_ p: [Double], _ U: Int) -> [Double] {
    // 
    // ADD CODE HERE
    // 
    let n = p.count
    let q = (Int(0)...n-1).map {(i: Int) -> Double in 
                  p[(i - (U + 1) + n) % n] * pUndershoot 
                + p[(i - U + n)       % n] * pExact 
                + p[(i - (U - 1) + n) % n] * pOvershoot
    }
    return q
}

// self-test /////////////////////////////////////////////////////////////////////
var q = p
for i in (0...motions.count-1) {
    q = move(sense(q, measurements[i]), motions[i])
}
print(q)

extension Sequence where Element: AdditiveArithmetic {
    /*
    ** sum any Numeric sequence
    */    
    func sum() -> Element {
        return reduce(.zero, +)
    }
}

extension Double {
    /*
    ** Test Double equality within a tolerance
    */
    func isApproximately(_ value: Double, _ tolerance: Double) -> Bool
    {
        return abs(self - value) <= abs(tolerance)
    }
}

extension Array where Element == Double {
    /*
    ** Test [Double] equality within a tolerance
    */
    func isApproximately(_ value: [Double], _ tolerance: Double) -> Bool
    {
        if(self.count != value.count) {
            return false
        }
        for i in (0...self.count - 1) {
            if(!self[i].isApproximately(value[i], tolerance)) {
                return false
            }
        }
        return true
    }
}


assert(q.sum().isApproximately(1.0, 0.0001))
assert(q.isApproximately([0.0788, 0.0753, 0.2247, 0.4329, 0.1882], 0.0001))
